text/0000-build-write-flag.md

```diff
@@ -0,0 +1,121 @@
+
+Unbeknownst to the buildpack author, python has a side-effect on execution where if it's installation directory is writable, it creates new cache objects or updates existing files. As a result, execution of buildpack #2 causes the content of the layers provided by buildpack #1 to change, causing the final layer imageID to change and hampers layer reproducibility and new layers have to be pushed out to the registry.
+
+This could easily be fixed if buildpack #1 only exposed its build layers as read-only to subsequent buildpacks or if it discarded the modifications made by subsequent buildpacks essentially making it "read-only" in the context of the build process.
+
+### Example use case for read and write build layers
+
+There are various use cases when you may want a common workspace that multiple buildpacks can collaborate on during the build process and this workspace needs to be cached, hence it can be a temporary directory.
+
+In such a case a buildpack could provide a layer that is marked as `build-writable` i.e. subsequent buildpacks can modify its contents and the final state of the layer at the end of the build process is what is exported out.
+
+One example might be an Android SDK buildpack that sets up ANDROID_SDK_ROOT (pointing to a layer), which a Gradle buildpack may later write into when gradle runs.
+
+Another example can be a CCache buildpack that sets up `CCACHE_DIR` (pointing to a layer), which a CMake buildpack may use for its build cache.
+
+## What is the expected outcome?
+
+The spec and lifecycle is modified to support the above use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+New layer types that dictate when the given layer directories are layerized. For layers marked as `build-write` to be `false`, they should be layerized immediately after the buildpack that provided the layer is finished with its build process. This way any modification made by the future buildpacks would be ignore while still making this layer available to future buildpacks.
+
+For layers that are marked as `build-write` as true, they should be layerized at the end of the entire build phase of the lifecycle so any changes made by subsequent buildpacks are present in the final app image.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle will have to layerize the layers marked `build-write` as `false` during the `build` phase of the lifecycle instead of the `export` phase.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Possibly mixing the concerns of the `build` and the `export` phases.
+- Added complexity of more flags for users.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+We could change the the layer so that it is owned by root or a different user between the build steps of different buildpacks so that subsequent buildpacks cannot modify the layers created by other buildpacks. This would require elevated privileges during the build phase of the lifecycle.
+
```

Comment:
@samj1912 not sure if you brought up an idea similar to this in WG but what if, as an alternative we introduce a new type of buildpack (name sure what to name it 😬) that runs twice during the build phase. It could run wherever it is placed in the order and generate layers, and run again after all the downstream buildpacks have executed, at which point it could write the layer content metadata. The example CCache buildpack mentioned above would be responsible for:
1. Scaffolding the `CCACHE_DIR` at the beginning of the build
2. Understanding the structure of the cache and recording its contents in the BOM at the end of the build

All normal buildpacks would create read-only layers.